### PR TITLE
tests: Fix flaky test for NLLB-MoE

### LINKED ISSUE #22464: Flaky test for NLLB-MoE Model
### System Info

Environment: Circle CI test_torch image, defined [here](https://github.com/huggingface/transformers/blob/c15f937581048fa28f77794c4f6a257e14d272cb/.circleci/create_circleci_config.py#L211)

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The test `tests/models/nllb_moe/test_modeling_nllb_moe.py::NllbMoeModelTest::test_decoder_model_past_with_large_inputs` is flaky and occasionally fails. Note: the tolerance on the test is also very low. 

Error cannot be deterministically reproduced. The test passes when I run it locally. 

An example of a failed run: https://app.circleci.com/pipelines/github/huggingface/transformers/60931/workflows/4642cd50-8d8c-4a6b-bea9-b077ff7400cf/jobs/747651

This came from this run: https://github.com/huggingface/transformers/runs/12389736536

And this unrelated PR: https://github.com/huggingface/transformers/pull/21855  

### Expected behavior

Tests pass on every run.

### PR DESCRIPTION
**Pull request by connor-henderson**
# What does this PR do?

Fixes #22464 (and added some light docs edits I happened to notice)

From my comment in the issue:
Looked into this and I think the flakiness is caused by the natural variability in the sparse MoE layers. Specifically that when they calculate which experts to use in the gating logic, they’re computing probabilities imperfectly for two different sets of inputs: one with prior inputs concatenated with the past key values and one with just the past key values. 

The test usually passes cause magnitude of the difference is usually likely to be small. Notably, when the vocab size is increased this pass rate goes up (and vice versa) since the increased representational capacity can help the model make more accurate decisions about which experts to use for each input. For example, increasing the vocab size in the config from its current 99 to 999 increases the pass rate from ~80% to ~95%.

I think this flakiness is inherent in the sparse layers, but if I understand right the point of the test is to check the decoder uses the past properly, so I edited the test to use dense layers and moved to the rtol down to 1e-3 to be in line with the other models’ version of this check. Wrote a loop to test this on a 1000 passes and they all passed.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker, @amyeroberts

### CODE REVIEW

File: docs/source/en/model_doc/nllb-moe.mdx [hunk 1]
```diff
@@ -43,9 +43,10 @@ This model was contributed by [Arthur Zucker](https://huggingface.co/ArtZucker).
 The original code can be found [here](https://github.com/facebookresearch/fairseq).
 
 ## Implementation differences with SwitchTransformers
-The biggest difference is the way the tokens are routed. NLLB-MoE uses a `top-2-gate` which means that blah blah blah blah. 
-In SwitchTransformers, once the masks are computed for each experts, we just index the current hidden_states with the routing mask, and feed the 
-correct tokens to the expert. However here, the implementation varies a lot as the fairseq repository used a different approach.
+The biggest difference is the way the tokens are routed. NLLB-MoE uses a `top-2-gate` which means that for each input, only the top two experts are selected based on the
```
Comment (amyeroberts) [inline docs/source/en/model_doc/nllb-moe.mdx:46]: 👀

**Comment by amyeroberts** [comment]
cc @ydshieh

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing this, for doing such a deep dive into the code, and for taking the time to write such detailed explanations here and on the issue 🙏 ❤️

File: docs/source/en/model_doc/nllb-moe.mdx [hunk 2]
```diff
@@ -43,9 +43,10 @@ This model was contributed by [Arthur Zucker](https://huggingface.co/ArtZucker).
 The original code can be found [here](https://github.com/facebookresearch/fairseq).
 
 ## Implementation differences with SwitchTransformers
-The biggest difference is the way the tokens are routed. NLLB-MoE uses a `top-2-gate` which means that blah blah blah blah. 
-In SwitchTransformers, once the masks are computed for each experts, we just index the current hidden_states with the routing mask, and feed the 
-correct tokens to the expert. However here, the implementation varies a lot as the fairseq repository used a different approach.
+The biggest difference is the way the tokens are routed. NLLB-MoE uses a `top-2-gate` which means that for each input, only the top two experts are selected based on the 
+highest predicted probabilities from the gating network, and the remaining experts are ignored. In SwitchTransformers, once the masks are computed for each expert, 
+we just index the current hidden_states with the routing mask, and feed the correct tokens to the expert. However here, the implementation varies a lot as the fairseq 
+repository used a different approach.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/nllb-moe.mdx:49]:
```suggestion
highest predicted probabilities from the gating network, and the remaining experts are ignored. In `SwitchTransformers`, only the top-1 probabilities are computes, which means that tokens have less probability of being forwarded. Moreover, if a token is not routed to any expert, `SwitchTransformers` still adds its unmodified hidden states (kind of like a residual connexion) while they are masked in `NLLB`'s top-2 routing mecanism. 
```

**Review by ArthurZucker — APPROVED** [review]
Thanks for tackling this 💪🏻

**Comment by connor-henderson** [comment]
Happy to!

**Pull request merged**